### PR TITLE
Add isEqual and isNotEqual utility

### LINKED ISSUE #3607: List: Add prop to allow deep compare on list content updates, and document pure/immutable defaults.
### Bug Report

- __Package version(s)__: 5.34.0 (current github code has this issue)
- __Browser and OS versions__: Any browser, any OS

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? Yes, but I don't what the fix should be

Requested priority: Normal

Products/sites affected: (if applicable)

#### Describe the issue:

There is this check in List.tsx's shouldComponentUpdate
```javascript
 if (newProps.items === this.props.items && ....)
```
The problem with this is this that this just compares the memory location that is pointed to by newProps.items, and this.props.items.  I was facing an issue where I have an array, and I went ahead and updated an entry within the same array, and passed it down to the list. Now since I didn't recreate the array, the memory location it points to is the same as before. They are pointing to the same array.

#### Expected behavior:

I am not sure what is the correct solution. As in my case, since the props objects are the same, there can never possibly be any check to distinguish between the two. When one newProps.items changes in any way, this.props.items changes along with it. The only thing that I was wondering is whether the props object should be deep extended and stored (not sure which component lifecycle could help here, or whether any would help at all), so that the issue of holding the same memory reference is avoided. This would enable correct comparisons between the old props and the new props.  Another option is to remove the check on this array altogether.

### Actionable items
(Copied from Cliff Koh's comment below)
For performance reason, the most sensible thing for the list to do here, by default, here is to behave as if the list item arrays are pure/immutable. The action item here is to document this expectation.

If there is a deep need to have a deep compare, it should at most be provided on a opt-in basis (like a isPure type of flag), or better still, allow users to provide the comparison function.

### PR DESCRIPTION
**Pull request by jordandrako**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3607
- [x] Include a change request file using `$ npm run change`

#### Description of changes

- Add `isEqual` (and `isNotEqual` function to utilities package that compares strings, numbers, arrays and objects more thoroughly than `array1 === array2` or `newProps.items === this.props.items`. 
- Add unit tests for new functions.

#### Focus areas to test

Test if this breaks List `shouldComponentUpdate` check and if it functions any better than a simple `===`.

### CODE REVIEW

File: packages/utilities/src/isEqual.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// tslint:disable-next-line no-any
+export const isEqual = (itemA: any, itemB: any): boolean => {
```
Comment (cliffkoh) [inline packages/utilities/src/isEqual.ts]: Please have JSDoc documenting this function :)

File: packages/utilities/src/isEqual.ts [hunk 2]
```diff
@@ -1,5 +1,9 @@
-// tslint:disable-next-line no-any
-export const isEqual = (itemA: any, itemB: any): boolean => {
+/**
+ * Checks if the first and second items are the NOT same recursively. Use for checking arrays and objects.
```
Comment (cliffkoh) [inline packages/utilities/src/isEqual.ts]: You have copy and paste issues unfortunately  :(

Comment (cliffkoh) [reply packages/utilities/src/isEqual.ts]: Also for `isNotEqual`, you probably want to fix the doc to say `are NOT the same, recursively.` instead.

Comment (jordandrako) [reply packages/utilities/src/isEqual.ts]: Whoops, thanks for catching that. Fixed my itchy copy+paste trigger finger.

**Comment by jordandrako** [comment]
Wow, sometimes I need to take break before switching branches and committing...

**Comment by cliffkoh** [comment]
"the NOT same" is back :P Would you like me to fix it for you?

**Comment by jordandrako** [comment]
I got it

Regards,

Jordan Janzen
UX Engineer
Aquent at Microsoft
Sent from Mobile

________________________________

**Comment by jordandrako** [comment]
Taking a break now, I swear...

**Comment by cliffkoh** [comment]
Don't stress about it :) Thanks for cleaning up!

**Comment by cliffkoh** [comment]
LGTM to merge. Just thinking out loud though, I wonder if it might have been better for Fabric to use say, lodash's isEqual rather than implementing our own version. This might possibly allow dedupe to take place in apps during the bundling phase for any apps that uses lodash as well.(https://npmcharts.com/compare/lodash,underscore,ramda). Then again, it is probably not a big deal since there's bunch of conditions needed before it can happen (lodash will need to resolve to the same version in both the app and Fabric UI React).

**Comment by jordandrako** [comment]
Are you suggesting adding lodash as a dependency in the utilities package? Wouldn't that drastically increase our bundle size?

**Comment by cliffkoh** [comment]
No, either

1) The full lodash package and rely on tree shaking to ensure a bundle size that is no greater than if we wrote it ourselves or

2) Just the lodash isEqual npm standalone package (which we won't benefit from dedupe however).

This is just me thinking out loud though - I am not recommending a switch at the moment :)

**Comment by jordandrako** [comment]
Something to think about, thanks for the suggestion.

**Pull request merged**